### ISSUE TITLE
AWS InstanceLimitExceeded & SSH Error No IP address

### ISSUE THREAD

**Issue by pnik073** [gc3-uzh-ch/elasticluster#537]
It is my first time using **elasticluster**. I use the following `slurm.conf` file to run SLURM on AWS (with custom AMI ami-174b2b6e - which works successfully on MPI cluster). However, I get the error `gc3.elasticluster[8686] ERROR Apparently, Amazon does not compute the RSA key fingerprint as we do! We cannot check if the uploaded keypair is correct!`.  Below I provide

1. The `slurm.conf`
2. The `elasticluster` command
3. Example of how I create my **AWS key-pair**

Can you give some feedback what could go wrong?

```
# slurm.conf

[setup/slurm]
provider=ansible
frontend_groups=slurm_master

compute_groups=slurm_worker,cuda

compute_var_allow_reboot=yes

global_var_slurm_taskplugin=task/cgroup
global_var_slurm_proctracktype=proctrack/cgroup


[cloud/amazon-eu-west-1]
provider=ec2_boto
ec2_url=https://ec2.eu-west-1.amazonaws.com
ec2_access_key=*****
ec2_secret_key=*****
ec2_region=eu-west-1

[login/ubuntu]
image_user=ubuntu
image_user_sudo=root
image_sudo=True
user_key_name=elasticluster
user_key_private=~/.ssh/elasticluster.pem
user_key_public=~/.ssh/elasticluster.pub

[cluster/slurm]
setup=slurm

cloud=amazon-eu-west-1
security_group=default

image_id=ami-174b2b6e
login=ubuntu

flavor=p2.xlarge

frontend_nodes=1
compute_nodes=2
ssh_to=frontend
```

```
vagrant@ubuntu-xenial:~/Documents$ elasticluster -c ~/Vagrant/quanticboost/vm/ec.d/slurm.conf start slurm
Starting cluster `slurm` with:
* 1 frontend nodes.
* 2 compute nodes.
(This may take a while...)
2018-03-25 23:43:33 ubuntu-xenial gc3.elasticluster[8686] ERROR Apparently, Amazon does not compute the RSA key fingerprint as we do! We cannot check if the uploaded keypair is correct!
2018-03-25 23:43:33 ubuntu-xenial gc3.elasticluster[8686] ERROR Apparently, Amazon does not compute the RSA key fingerprint as we do! We cannot check if the uploaded keypair is correct!
2018-03-25 23:43:33 ubuntu-xenial gc3.elasticluster[8686] ERROR Apparently, Amazon does not compute the RSA key fingerprint as we do! We cannot check if the uploaded keypair is correct!
2018-03-25 23:43:33 ubuntu-xenial gc3.elasticluster[8686] ERROR Error starting instance: EC2ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>InstanceLimitExceeded</Code><Message>You have requested more instances (2) than your current instance limit of 1 allows for the specified instance type. Please visit http://aws.amazon.com/contact-us/ec2-request to request an adjustment to this limit.</Message></Error></Errors><RequestID>d1438ca3-2a3f-4c96-809e-94e5abb5fc45</RequestID></Response>
2018-03-25 23:43:33 ubuntu-xenial gc3.elasticluster[8686] ERROR Could not start node `frontend001`: EC2ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>InstanceLimitExceeded</Code><Message>You have requested more instances (2) than your current instance limit of 1 allows for the specified instance type. Please visit http://aws.amazon.com/contact-us/ec2-request to request an adjustment to this limit.</Message></Error></Errors><RequestID>d1438ca3-2a3f-4c96-809e-94e5abb5fc45</RequestID></Response> -- <class 'elasticluster.exceptions.InstanceError'>
Traceback (most recent call last):
  File "/home/vagrant/elasticluster/src/elasticluster/cluster.py", line 540, in _start_node
    node.start()
  File "/home/vagrant/elasticluster/src/elasticluster/cluster.py", line 1183, in start
    **self.extra)
  File "/home/vagrant/elasticluster/src/elasticluster/providers/ec2_boto.py", line 279, in start_instance
    raise InstanceError(ex)
InstanceError: EC2ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>InstanceLimitExceeded</Code><Message>You have requested more instances (2) than your current instance limit of 1 allows for the specified instance type. Please visit http://aws.amazon.com/contact-us/ec2-request to request an adjustment to this limit.</Message></Error></Errors><RequestID>d1438ca3-2a3f-4c96-809e-94e5abb5fc45</RequestID></Response>
2018-03-25 23:43:34 ubuntu-xenial gc3.elasticluster[8686] ERROR Error starting instance: EC2ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>InstanceLimitExceeded</Code><Message>You have requested more instances (2) than your current instance limit of 1 allows for the specified instance type. Please visit http://aws.amazon.com/contact-us/ec2-request to request an adjustment to this limit.</Message></Error></Errors><RequestID>b1bd0812-0549-45dc-92df-d1b598b9246d</RequestID></Response>
2018-03-25 23:43:34 ubuntu-xenial gc3.elasticluster[8686] ERROR Could not start node `compute001`: EC2ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>InstanceLimitExceeded</Code><Message>You have requested more instances (2) than your current instance limit of 1 allows for the specified instance type. Please visit http://aws.amazon.com/contact-us/ec2-request to request an adjustment to this limit.</Message></Error></Errors><RequestID>b1bd0812-0549-45dc-92df-d1b598b9246d</RequestID></Response> -- <class 'elasticluster.exceptions.InstanceError'>
Traceback (most recent call last):
  File "/home/vagrant/elasticluster/src/elasticluster/cluster.py", line 540, in _start_node
    node.start()
  File "/home/vagrant/elasticluster/src/elasticluster/cluster.py", line 1183, in start
    **self.extra)
  File "/home/vagrant/elasticluster/src/elasticluster/providers/ec2_boto.py", line 279, in start_instance
    raise InstanceError(ex)
InstanceError: EC2ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>InstanceLimitExceeded</Code><Message>You have requested more instances (2) than your current instance limit of 1 allows for the specified instance type. Please visit http://aws.amazon.com/contact-us/ec2-request to request an adjustment to this limit.</Message></Error></Errors><RequestID>b1bd0812-0549-45dc-92df-d1b598b9246d</RequestID></Response>
```

```
vagrant@ubuntu-xenial:~$ sudo pip install awscli

vagrant@ubuntu-xenial:~$ aws configure
AWS Access Key ID [None]: ***
AWS Secret Access Key [None]: ***
Default region name [None]: eu-west-1
Default output format [None]: json

vagrant@ubuntu-xenial:~$ aws ec2 create-key-pair --key-name elasticluster --query 'KeyMaterial' --output text > ~/.ssh/elasticluster.pem
vagrant@ubuntu-xenial:~$ ssh-keygen -y -f ~/.ssh/elasticluster.pem > ~/.ssh/elasticluster.pub
vagrant@ubuntu-xenial:~$ vim ~/.ssh/elasticluster.pub
vagrant@ubuntu-xenial:~$ chmod 400 /home/vagrant/.ssh/elasticluster.pub 
```

**Comment by riccardomurri** [comment]
Hello, @pnik073 thanks for using ElastiCluster! This error is harmless::

    ERROR Apparently, Amazon does not compute the RSA key fingerprint as we do! We cannot check if the uploaded keypair is correct!

It is just a warning that ElastiCluster cannot verify the RSA key
fingerprint so it cannot be sure that it's going to work eventually.
The issue is actually solved in PR #536 which will be merged in the
next days.

What is stopping you from starting a cluster is *this* error instead::

    <Code>InstanceLimitExceeded</Code>
    <Message>
      You have requested more instances (2) than your current
      instance limit of 1 allows for the specified instance type.
      Please visit http://aws.amazon.com/contact-us/ec2-request
      to request an adjustment to this limit.
    </Message>

You're trying to start a cluster with 2x compute nodes of type
`p2.xlarge` (GPU-enabled) but your AWS quota only allows 1x; request
an increase or use compute nodes without a GPU, which are not
restricted by AWS.

**Comment by pnik073** [comment]
Hi @riccardomurri thanks for the hint. I got it running with t2.medium for testing purposes and I get the following **SSH Error** (similar to #490), see below for details. My AMI is Ubuntu Xenial (I must to use that version as my C++/cuda libs are based on its provisioning).

*What do you suggest on that?*

**DETAILS:**

> I use the GIT version of the `elasticluster` as the `pip install` did not for me.
> Also my `pip list` shows that I use boto (2.48.0) and botocore (1.9.16).
> And my AWS SG is `default` with the following settings (I can SSH other nodes)

|Type | Protocol | Port Range | Source |
|:-- | :-- | :-- | :-- | 
|All traffic | All | All | 0.0.0.0/0 |  
|SSH | TCP | 22 | 0.0.0.0/0 |  
|Custom TCP Rule | TCP | 5900 - 5950 | 0.0.0.0/0|

**Error while provisioning**

```
Configuring the cluster ...
(this too may take a while)
2018-03-26 23:57:16 ubuntu-xenial gc3.elasticluster[5829] WARNING Ignoring node `frontend001`: No IP address.
2018-03-26 23:57:16 ubuntu-xenial gc3.elasticluster[5829] WARNING Ignoring node `compute001`: No IP address.
2018-03-26 23:57:16 ubuntu-xenial gc3.elasticluster[5829] WARNING Ignoring node `compute002`: No IP address.
```

**Error while SSHing**

```
$ elasticluster -c ~/Vagrant/quanticboost/vm/ec.d/slurm.conf ssh slurm
2018-03-27 00:00:46 ubuntu-xenial gc3.elasticluster[5977] ERROR No IP address known for node frontend001
$ sudo ssh -i ~/.ssh/elasticluster.pem root@ec2-54-154-1-240.eu-west-1.compute.amazonaws.com
ssh: connect to host ec2-54-154-1-240.eu-west-1.compute.amazonaws.com port 22: Connection timed out
```

**Comment by riccardomurri** [comment]
These lines suggest that the cluster has not been even started, but
it's impossible to say why without access to the creation logs:

> 2018-03-26 23:57:16 ubuntu-xenial gc3.elasticluster[5829] WARNING Ignoring node `frontend001`: No IP address.
> 2018-03-26 23:57:16 ubuntu-xenial gc3.elasticluster[5829] WARNING Ignoring node `compute001`: No IP address.
> 2018-03-26 23:57:16 ubuntu-xenial gc3.elasticluster[5829] WARNING Ignoring node `compute002`: No IP address.

Do you see instances `slurm-frontend001`, `slurm-compute001`, `slurm-compute002` on the AWS console?
If yes, I would suggest that you terminate them from the console and start over.

Before running `elasticluster start` again, please delete the current cluster::

        elasticluster stop -y slurm
        rm -f -v ~/.elasticluster/storage/slurm*

Then run again `elasticluster -c ... -v start slurm`; if it still does
not work, please post (or send me by email) the entire output.

**Comment by pnik073** [comment]
@riccardomurri I summarize my current findings:

* I stop and start the nodes from AWS console. No success.
* I have re-tested the PEM file by launching a new instance (from EC2 console). The SSH with the PEM works well.
* After `elasticluster -c ec.d/slurm.conf stop slurm` the `~/.elasticluster/storage/` is always empty. No need to `rm` content.
* When I `elasticluster -c ec.d/slurm.conf start slurm` I see the nodes on AWS. The `~/.elasticluster/storage/slurm.known_hosts` is empty and the `~/.elasticluster/storage/slurm.yaml` is given below.

> **Any further suggestions for troubleshooting?**

```
vagrant@ubuntu-xenial:~$ cat .elasticluster/storage/slurm.yaml 
extra:
    accelerator_count: 0
    allow_project_ssh_keys: true
    compute_nodes: '2'
    flavor: t2.medium
    frontend_nodes: '1'
    image_id: ami-174b2b6e
    login:
        image_sudo: true
        image_user: ubuntu
        image_user_sudo: root
        image_userdata: ''
        user_key_name: elasticluster
        user_key_private: /home/vagrant/.ssh/elasticluster.pem
        user_key_public: /home/vagrant/.ssh/elasticluster.pub
    security_group: default
name: slurm
nodes:
    compute:
    -   cluster_name: slurm
        extra:
            accelerator_count: 0
            login: ubuntu
            min_num: 2
        flavor: t2.medium
        image_id: ami-174b2b6e
        image_user: ubuntu
        image_userdata: ''
        instance_id: i-0671e0fe8e17814c3
        ips:
        - 34.244.37.146
        - 172.31.32.100
        kind: compute
        name: compute001
        preferred_ip: null
        security_group: default
        user_key_name: elasticluster
        user_key_private: /home/vagrant/.ssh/elasticluster.pem
        user_key_public: /home/vagrant/.ssh/elasticluster.pub
    -   cluster_name: slurm
        extra:
            accelerator_count: 0
            login: ubuntu
            min_num: 2
        flavor: t2.medium
        image_id: ami-174b2b6e
        image_user: ubuntu
        image_userdata: ''
        instance_id: i-098ea184d1b531cf9
        ips:
        - 34.241.184.236
        - 172.31.40.54
        kind: compute
        name: compute002
        preferred_ip: null
        security_group: default
        user_key_name: elasticluster
        user_key_private: /home/vagrant/.ssh/elasticluster.pem
        user_key_public: /home/vagrant/.ssh/elasticluster.pub
    frontend:
    -   cluster_name: slurm
        extra:
            accelerator_count: 0
            login: ubuntu
            min_num: 1
        flavor: t2.medium
        image_id: ami-174b2b6e
        image_user: ubuntu
        image_userdata: ''
        instance_id: i-073bf36cfd56acedb
        ips:
        - 34.241.10.140
        - 172.31.45.140
        kind: frontend
        name: frontend001
        preferred_ip: null
        security_group: default
        user_key_name: elasticluster
        user_key_private: /home/vagrant/.ssh/elasticluster.pem
        user_key_public: /home/vagrant/.ssh/elasticluster.pub
ssh_probe_timeout: 5
ssh_to: frontend
start_timeout: 600
template: slurm
thread_pool_max_size: 10
user_key_name: elasticluster
user_key_private: /home/vagrant/.ssh/elasticluster.pem
user_key_public: /home/vagrant/.ssh/elasticluster.pub
```

**Comment by riccardomurri** [comment]
Thanks for posting this information -- things look much clearer now.

I think this is an issue with the "default" security group in your AWS account, which does not allow SSH connections from the computer where you're running ElastiCluster. Either add a rule to the "default" security group to allow inbound SSH, or run ElastiCluster from a VM instance in AWS (within the same security group as the cluster).

**Comment by pnik073** [comment]
The default group for me is:

Type | Protocol | Port Range | Source
-- | -- | -- | --
All traffic | All | All | 0.0.0.0/0
SSH | TCP | 22 | 0.0.0.0/0
Custom TCP Rule | TCP | 5900 - 5950 | 0.0.0.0/0

and I am able to SSH from my dev machine to AWS for any other node. Can it be something else?

**Comment by riccardomurri** [comment]
> I am able to SSH from my dev machine to AWS for any other node. Can it be something else?

Well, the log file you posted looks quite explicit:

> 2018-03-27 21:51:03 ubuntu-xenial gc3.elasticluster[8279] INFO Checking SSH connection to nodes (timeout: 600 seconds) ...
> 2018-03-27 22:01:03 ubuntu-xenial gc3.elasticluster[8279] ERROR Some nodes of the cluster were unreachable within the given 600-seconds timeout: compute002, frontend001, compute001

In between these two log lines, all ElastiCluster does is trying to
SSH to all IP addresses of each VM, until one finally responds. (You
can see that if you increase the logging verbosity to DEBUG with
`elasticluster -vvv`)

It's of course possible that the VMs do not run the SSH daemon, or
that they only do so on the private IP address, or that they don't
configure networking properly -- but any of these would make the VMs
basically unusable on AWS, so I presume this cannot be the issue...

**Comment by riccardomurri** [comment]
(Please don't edit comments you already posted, add new ones... GitHub does not send notification for edits, so I won't see the changes unless I reload the page and go re-read your text.)

**Comment by riccardomurri** [comment]
What's `launch-wizard-7` ?

**Comment by pnik073** [comment]
@riccardomurri I can successfully SSH using the `elasticluster ssh` command! Now, I hit on another challenge with respect to `SSL certificate for github.com:443`. I found some [bug report](https://github.com/ansible/ansible/issues/33417) related to `get_url`, where some people suggest to use `validate_certs: no` if the error persists. Any ideas on that? 

PS: I appreciate your responsiveness.

```
TASK [lmod : Download sources] ***************************************************************************************
fatal: [frontend001]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to validate the SSL certificate for github.com:443. Make sure your managed systems have a valid CA certificate installed. You can use validate_certs=False if you do not need to confirm the servers identity but this is unsafe and not recommended. Paths checked for this platform: /etc/ssl/certs, /etc/pki/ca-trust/extracted/pem, /etc/pki/tls/certs, /usr/share/ca-certificates/cacert.org, /etc/ansible. The exception msg was: (\"bad handshake: Error([('SSL routines', 'ssl3_read_bytes', 'tlsv1 alert protocol version')],)\",)."}

PLAY RECAP ***********************************************************************************************************
compute001                 : ok=47   changed=30   unreachable=0    failed=0   
compute002                 : ok=47   changed=30   unreachable=0    failed=0   
frontend001                : ok=71   changed=46   unreachable=0    failed=1   

2018-03-28 22:59:16 ubuntu-xenial gc3.elasticluster[10571] ERROR Command `ansible-playbook --private-key=/home/vagrant/.ssh/elasticluster.pem /home/vagrant/elasticluster/src/elasticluster/share/playbooks/site.yml --inventory=/home/vagrant/.elasticluster/storage/slurm.inventory --become --become-user=root -e elasticluster_output_dir=/tmp/elasticluster._bksmv.d` failed with exit code 2.
2018-03-28 22:59:16 ubuntu-xenial gc3.elasticluster[10571] ERROR Cannot find the status report file.
2018-03-28 22:59:16 ubuntu-xenial gc3.elasticluster[10571] WARNING The cluster has likely *not* been configured correctly. You may need to re-run `elasticluster setup`.
2018-03-28 22:59:16 ubuntu-xenial gc3.elasticluster[10571] WARNING Cluster `slurm` not yet configured. Please, re-run `elasticluster setup slurm` and/or check your configuration

WARNING: YOUR CLUSTER `slurm` IS NOT READY YET!
```

**Comment by riccardomurri** [comment]
If it's not a transient problem (i.e., re-running `elasticluster setup slurm` does not fix it), then you can implement the fix yourself by editing file `/home/vagrant/elasticluster/src/elasticluster/share/playbooks/roles/lmod/tasks/build.yml` and add this at line 70 (i.e., the line after `get_url:`)::

    validate: no

Note that the `v` of `validate` must be exactly aligned with the `u` of `url:` in the line that follows.
(This is indeed the workaround suggested in the bug report that you linked -- nice find!)

If that works, the question would be: why isn't the certificate from GitHub.com not validating on your system?  What operating system are you using (for running ElastiCluster)?  Any chance you could use the  [Docker image](http://elasticluster.readthedocs.io/en/latest/install.html#quickstart) instead?

(As a side note: please open new issue reports for unrelated problems: now we have an SSL problem under an AWS title and people coming from the linked Ansible `get_url` bug report will be confused...)

**Comment by riccardomurri** [comment]
Conversation continues on #539

**Comment by riccardomurri** [comment]
(Many thanks for moving the unrelated part to a different issue report, with full details!  I appreciate the effort to give the full context!)
